### PR TITLE
Added select tool to image editor 

### PR DESCRIPTION
**Pull request by Danielku15**
This pull request adds a new select tool to the image editor as proposed in #3769 

Tasks: 
- [x] Add select tool
- [x] Change escape key behavior
- [x] Update select tool icon
- [x] Add new setting that will allows the user to switch between the select modes: "Keep select tool active after shape selection", "Select shape tool after shape selection".
- [x] Add new setting that will allows the user to switch between the draw modes: "Keep current draw tool active after shape drawing", "Switch to select tool after shape creation"
- [x] Fix Crop Tool and use correct icon

### CODE REVIEW

File: ShareX.ScreenCaptureLib/Shapes/ShapeManager.cs [hunk 1]
```diff
@@ -90,7 +93,10 @@ private set
                     ClearTools();
                 }
 
-                DeselectCurrentShape();
+                if(previousTool != ShapeType.ToolSelect && currentTool != ShapeType.ToolSelect)
```
Comment (Jaex) [inline ShareX.ScreenCaptureLib/Shapes/ShapeManager.cs]: Why there is no space after `if` here?

**Comment by Jaex** [comment]
I wonder is it good idea to use rectangle selection icon like this in Photoshop:



when ShareX not supports selecting multiple shapes with rectangle selection.

And in that case also shortcut of <kbd>M</kbd> becomes invalid too.

**Comment by Jaex** [comment]
In image editor mode when "Switch to selection too after shape drawing" is selected, crop tool not works.

**Comment by Danielku15** [comment]
> I wonder is it good idea to use rectangle selection icon like this in Photoshop:
> 
> 
> 
> when ShareX not supports selecting multiple shapes with rectangle selection.

Any icon you would prefer? Initially I had the cursor icon in mind, but this is already used for drawing a cursor. Beside that also the icons image-map and layer-select would fit quite good. I like the move tool icons from paint.net but I could not find some icon like this in the fugue icon set: 
image

> And in that case also shortcut of M becomes wrong too.

Depends on how you look at it 😉 . In Photoshop the <kbd>M</kbd> stands for "Marquee" while in Paint.net it is rather "Move". <kbd>S</kbd> is already used for "Speech-Bubble" and can therefore not be used for "Select" anymore. Any proposal that you have as alternative? 

> In image editor mode when "Switch to selection too after shape drawing" is selected, crop tool not works.
> Also selecting "select tool" after crop tool drawing have issue too.

I will have a look on that. Likely I will limit this "feature/option" to the actual drawing tools. Then it should not impact the crop tool anymore. But first I will check how it actually affects it. A fix will come soon.

**Comment by Jaex** [comment]
Using cursor icon for select tool and finding another icon for cursor drawing tool gonna be more easier I guess.

It is fine then if "M" means move.

**Comment by Jaex** [comment]
stamp-cursor
I made this icon for cursor tool now. So current cursor icon can be used for select tool.

**Comment by Danielku15** [comment]
I made some more adjustments and now the crop tool is working fine again. Also after cropping the setting is considered and the "select tool" is activated. Let me know if you discover more

File: ShareX.ScreenCaptureLib/RegionCaptureOptions.cs [hunk 2]
```diff
@@ -73,11 +73,14 @@ public class RegionCaptureOptions
         public bool MenuCollapsed = false;
         public Point MenuPosition = Point.Empty;
 
+        public bool SwitchToDrawingToolAfterSelection = true;
```
Comment (Jaex) [inline ShareX.ScreenCaptureLib/RegionCaptureOptions.cs]: I think making this option `false` by default would make more sense for users.

**Comment by Jaex** [comment]
Great work, thanks for implementing it!

**Comment by Jaex** [comment]
I noticed now this issue I reported:

> Also selecting "select tool" after crop tool drawing have issue too.

Is still exists.

**Comment by Danielku15** [comment]
> Is still exists.

I made various tests and it worked fine for me? Can I assume, as you merged the PR, that the issue disappeared? 

Also there are some other remarks. Will you take care of this decision or do you want me to follow up on anything? 😉

**Comment by Jaex** [comment]
I just forget to test it again before merge :D I assumed you fixed it.

This is the issue:



Yea I was asking your opinion about should that option be false by default.

**Comment by Jaex** [comment]
When you select shape with select tool and then change to any drawing tool then selection still stays active:



But if you change from one drawing tool to another tool then as intended it unselects previous shape.

**Comment by Danielku15** [comment]
Okay, thanks for the info. I will try to fix the remaining 3 points and will file another PR:
- [ ] [Bug] Crop-Tool Selection must be cleared when switching to the select tool 
- [ ] [Bug] Select-Tool Selection must be cleared when switching to any other tool 
- [ ] [Change] Make the auto-selection of tools disabled by default

**Comment by Jaex** [comment]
Found another issue which happens if you select resizeable shape first then not resizeable:

**Comment by Danielku15** [comment]
Wow, looks like there are quite a lot of individual tool checks scattered across the code that rely that the current tool matches the shape. I will have a closer look if I can get this cleaned.

**Comment by Jaex** [comment]
Hi @Danielku15, this release a lot of people were reporting that their annotation shapes border size is resetting to 0 and their shapes becomes invisible. Recently I noticed in here #4418 that issue is actually related to "Select tool".

Reproduce steps:

1. Select any text tool and set border size to 0 then add text to screen.
2. Use "Select tool" and select the text.
3. Switch to "Line tool".

You will see that line tool border size changed to 0 which is previously selected shapes border size with cursor tool. When you have time can you check this issue?

**Comment by Jaex** [comment]
I fixed the issue now.

**Pull request merged**